### PR TITLE
Numba montecarlo

### PR DESCRIPTION
**Pull request by wkerzendorf**
## Things to check if they are done ##

Things to do 

 - [x] line scattering
 - [x] electron scattering
 - [x] virtual packets
 - [x] macroatom

- [x] packet logging (last_packet_interaction)

Line estimators are slowing the process down - maybe a useful optimization strategy

## Final Testing ##

There are 4 code versions. There is `upstream/master` there is `numba_montecarlo` both of which have different seeding strategies and will never be numerically equal (those we will summarize as main). There are two versions that have additional code that will allow logging - those we will call `c_compare` and `numba_compare` (and will call the compare branch). 
 - [x] virtual - only compare - sica
 - [x] virtual - only compare - ddc10
 - [ ] integrated - main only - sica
 - [ ] integrated - main only - ddc10
 - [x] estimators - compare only -sica
 - [x] estimators - compare only -ddc10
 - [x] iterations - (compare/main) -sica
 - [x] iterations - (compare/main) -ddc10

 - [x] Need to fix negative boundary distance issue for vpackets: specific case occurs when a packet reaches a boundary with a close line so distance_trace_line = 0 and distance_boundary can be negative (shell ID does not advance)
The models that shall be run is low velocity Si/Ca and DDC10. 

The last step is to get numba to produce a new refdata and look at this in the notebook. 

If all these models run then we agree that numba montecarlo can be merged. 

## Merging Strategy ##

1. talk to TARDIS core 
2. if all tests pass - update the version number to 4.0 with the merged numba montecarlo
3. merge

## Post-merge ##

Start writing unit tests for the numba version
Improve speed.

### CODE REVIEW

**Comment by wkerzendorf** [comment]
There is a sensible discussion about setting `r` when doing a `move_packet_across_shell_boundary` - do we move like we normally do - or just set `r` to the shell boundary??

File: tardis/montecarlo/montecarlo_numba/__init__.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+from llvmlite import binding
+binding.set_option("tmp", "-non-global-value-max-name-size=2048")
```
Comment (stuartarchibald) [inline tardis/montecarlo/montecarlo_numba/__init__.py]: Wondering what this is for? IIRC this patch against LLVM "fixed", well, prevented, it from being an issue https://github.com/numba/llvmlite/blob/master/conda-recipes/0001-Revert-Limit-size-of-non-GlobalValue-name.patch .

Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/__init__.py]: good to know. Thanks - haven't worked on it in a while.

**Comment by shreyansjain012** [comment]
@wkerzendorf - Can you please elaborate on the changes required in this branch and how can I get started with them.

**Comment by andrewfullard** [comment]
Don't want to match vpackets exactly- prefer summary statitics comparison.

- [ ] Need to fix negative boundary distance issue for vpackets
- [ ] Need to fix formal integral

**Comment by andrewfullard** [comment]
image
Si Ca atmosphere 10000 km/s, 10 day. Real packets only. Numba code 2.2x slower.

File: tardis/montecarlo/montecarlo_numba/tests/test_packet.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    ['packet_params', 'expected_params'],
+    [({'mu': 0.3, 'r': 7.5e14},
+      {'d_boundary': 259376919351035.88}),
+
+     ({'mu': -.3, 'r': 7.5e13},
+      {'d_boundary': -664987228972291.5}),
+
+     ({'mu': -.3, 'r': 7.5e14},
+      {'d_boundary': 709376919351035.9})]
+)
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/test_packet.py:61]: @andrewfullard @Rodot- we should use more easy to understand mus and nus here. I would say put it in the middle of the cell and go mu=1, -1 - and then mu=0 and use the secant calculation to see if thins calculates the thing. Does it also not return if it goes out or in (`d_shell`)?

File: tardis/montecarlo/montecarlo_numba/tests/test_packet.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    assert_almost_equal(obtained, expected)
+
+
+def test_get_random_mu():
+    """
+    Ensure that different calls results
+    """
+    output1 = r_packet.get_random_mu()
+    output2 = r_packet.get_random_mu()
+    assert output1 != output2
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/test_packet.py:165]: like that :-D

Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/tests/test_packet.py:165]: is there a way to init the seed everytime for a module.

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+    return distance
+
+@njit(**njit_dict)
+def calculate_distance_electron(electron_density, tau_event):
+    # add full_relativity here
+    return tau_event / (electron_density * numba_config.SIGMA_THOMSON)
+
+@njit(**njit_dict)
+def calculate_tau_electron(electron_density, distance):
+    return electron_density * numba_config.SIGMA_THOMSON * distance
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:163]:
```suggestion
    """
    Calculate tau for Thomson scattering
    
    Parameters:
    ------------
    
    electron_density: float
    distance: float
    """
    return electron_density * numba_config.SIGMA_THOMSON * distance
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+
+@njit(**njit_dict)
+def get_doppler_factor_full_relativity(mu, beta):
+    return (1.0 - mu * beta) / math.sqrt(1 - beta * beta)
+
+
+
+@njit(**njit_dict)
+def get_inverse_doppler_factor(r, mu, time_explosion):
+    inv_c = 1 / C_SPEED_OF_LIGHT
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:188]:
```suggestion
    """
    Calculate doppler factor for frame transformation
    
    Parameters:
    ------------
    
    r: float
    mu: float
    time_explosion: float 
    """
    inv_c = 1 / C_SPEED_OF_LIGHT
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+    """
+    Function to update the line estimators
+
+    Parameters
+    ----------
+    estimators
+    r_packet
+    cur_line_id
+    distance_trace
+    time_explosion
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:222]:
```suggestion
    estimators: Estimators 
    r_packet: RPacket
    cur_line_id: int
    distance_trace: float
    time_explosion: float
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+def calc_packet_energy(r_packet, distance_trace, time_explosion):
+    doppler_factor = 1.0 - ((distance_trace + r_packet.mu * r_packet.r) /
+                            (time_explosion * C_SPEED_OF_LIGHT))
+    energy = r_packet.energy * doppler_factor
+    return energy
+
+@njit(**njit_dict)
+def trace_packet(r_packet, numba_model, numba_plasma, estimators):
+    """
+
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:260]:
```suggestion
    Traces the RPacket through the ejecta and stops when an interaction happens (heart of the calculation)
    
    r_packet: RPacket
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+    #r_packet.next_line_id = cur_line_id
+
+    r_packet.last_interaction_type = interaction_type
+    
+    return distance, interaction_type, delta_shell
+
+
+@njit(**njit_dict)
+def move_r_packet(r_packet, distance, time_explosion, numba_estimator):
+    """Move packet a distance and recalculate the new angle mu
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:391]:
```suggestion
    """
    Move packet a distance and recalculate the new angle mu
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+            distance = distance * doppler_factor
+            set_estimators_full_relativity(r_packet,
+                                           distance,
+                                           numba_estimator,
+                                           comov_nu,
+                                           comov_energy,
+                                           doppler_factor)
+@njit(**njit_dict)
+def set_estimators(r_packet, distance, numba_estimator, comov_nu, comov_energy):
+    numba_estimator.j_estimator[r_packet.current_shell_id] += (
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:437]:
```suggestion
   """
   Updating the estimators
   """
    numba_estimator.j_estimator[r_packet.current_shell_id] += (
```

File: tardis/montecarlo/montecarlo_numba/r_packet.py [hunk 10]
```diff
@@ -0,0 +1,534 @@
...
+    doppler_factor = get_doppler_factor(r_packet.r,
+                                        r_packet.mu,
+                                        time_explosion,)
+    r_packet.nu = numba_plasma.line_list_nu[emission_line_id] / doppler_factor
+    r_packet.next_line_id = emission_line_id + 1
+    if montecarlo_configuration.full_relativity:
+        r_packet.mu = angle_aberration_CMF_to_LF(
+            r_packet,
+            time_explosion
+            )
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/r_packet.py:478]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/tests/conftest.py [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+from tardis.simulation import Simulation
+from tardis.montecarlo.montecarlo_numba import RPacket, PacketCollection
+
+
+from tardis.montecarlo.montecarlo_numba.numba_interface import (
+    numba_plasma_initialize, NumbaModel, Estimators, VPacketCollection)
+
+#from tardis.montecarlo.montecarlo_numba.r_packet import (trace_packet, get_doppler_factor, get_doppler_factor_partial_relativity)
+
+#from tardis.montecarlo import montecarlo_configuration
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/conftest.py:16]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/tests/test_packet.py [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+    ['distance_trace', 'time_explosion'],
+    [(0, 1),
+     (1, 1),
+     (1, 1e9)]
+)
+def test_packet_energy_limit_one(packet, distance_trace, time_explosion):
+    initial_energy = packet.energy
+    new_energy = r_packet.calc_packet_energy(packet, distance_trace, time_explosion)
+    assert_almost_equal(new_energy, initial_energy)
+"""
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/test_packet.py:329]:
```suggestion
```

File: tardis/montecarlo/montecarlo_numba/tests/test_single_packet_loop.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+    npt.assert_almost_equal(r_packet.mu, 0.9611146425440562)
+    npt.assert_almost_equal(r_packet.energy, 0.10327717505563379)
+
+@pytest.mark.xfail(reason='To be implemented')
+def test_set_packet_props_partial_relativity():
+    assert False
+
+@pytest.mark.xfail(reason='To be implemented')
+def test_set_packet_props_full_relativity():
+    assert False
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/test_single_packet_loop.py:41]:
```suggestion
    assert False
    
```

File: tardis/montecarlo/montecarlo_numba/tests/test_vpacket.py [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        mu = 0.4916053094346575,
+        energy = 2.474533071386993e-07,
+        index = 3,
+        is_close_line = True,
+        current_shell_id = 0,
+        next_line_id = 5495
+    )
+
+def test_trace_bad_vpacket(broken_packet, verysimple_numba_model, verysimple_numba_plasma):
+    vpacket.trace_vpacket(broken_packet, verysimple_numba_model, verysimple_numba_plasma)
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/tests/test_vpacket.py:102]:
```suggestion
    vpacket.trace_vpacket(broken_packet, verysimple_numba_model, verysimple_numba_plasma)
    
```

File: tardis/montecarlo/montecarlo_numba/vpacket.py [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+        self.current_shell_id = current_shell_id
+        self.next_line_id = next_line_id
+        self.status = PacketStatus.IN_PROCESS
+        self.index = index
+        self.is_close_line = is_close_line
+
+
+@njit(**njit_dict)
+def trace_vpacket_within_shell(v_packet, numba_model, numba_plasma):
+
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/vpacket.py:43]:
```suggestion
    """
    Trace VPacket within one shell (relatively simple operation)
    """
```

File: tardis/montecarlo/montecarlo_numba/vpacket.py [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+            )
+
+        """
+        assert vpacket_collection.idx < montecarlo_configuration.temporary_v_packet_bins, \
+            "ERROR: vpacket collection overflowed!"
+             
+        vpacket_collection.nus[vpacket_collection.idx] = v_packet.nu
+        vpacket_collection.energies[vpacket_collection.idx] = v_packet.energy
+        vpacket_collection.idx += 1
+        """
```
Comment (wkerzendorf) [inline tardis/montecarlo/montecarlo_numba/vpacket.py:243]:
```suggestion
```

File: tardis/scripts/debug/run_numba_single.run.xml [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    <option name="SCRIPT_NAME" value="$PROJECT_DIR$/../../../numba/run_numba_single.py" />
+    <option name="PARAMETERS" value="&quot;seed=46&quot;" />
+    <option name="SHOW_COMMAND_LINE" value="true" />
+    <option name="EMULATE_TERMINAL" value="false" />
+    <option name="MODULE_MODE" value="false" />
+    <option name="REDIRECT_INPUT" value="false" />
+    <option name="INPUT_FILE" value="" />
+    <method v="2" />
+  </configuration>
+</component>
```
Comment (wkerzendorf) [inline tardis/scripts/debug/run_numba_single.run.xml:25]:
```suggestion
</component>

```

**Pull request merged**